aptos-move/mvhashmap/src/lib.rs

```diff
@@ -2,304 +2,126 @@
-    pub(crate) contains_delta: bool,
-}
-
-impl<V: TransactionWrite> VersionedValue<V> {
-    pub fn new() -> Self {
-        Self {
-            versioned_map: BTreeMap::new(),
-            contains_delta: false,
-        }
-    }
-}
-
-impl<V: TransactionWrite> Default for VersionedValue<V> {
-    fn default() -> Self {
-        VersionedValue::new()
-    }
-}
-
 /// Main multi-version data-structure used by threads to read/write during parallel
-/// execution. Maps each access path to an interal BTreeMap that contains the indices
-/// of transactions that write at the given access path alongside the corresponding
-/// entries of WriteCell type.
+/// execution.
 ///
 /// Concurrency is managed by DashMap, i.e. when a method accesses a BTreeMap at a
 /// given key, it holds exclusive access and doesn't need to explicitly synchronize
 /// with other reader/writers.
-pub struct MVHashMap<K, V> {
-    data: DashMap<K, VersionedValue<V>>,
-    delta_keys: Mutex<Vec<K>>,
-}
-
-/// Returned as Err(..) when failed to read from the multi-version data-structure.
-#[derive(Debug, PartialEq, Eq)]
-pub enum MVHashMapError {
-    /// No prior entry is found.
-    NotFound,
-    /// Read resulted in an unresolved delta value.
-    Unresolved(DeltaOp),
-    /// A dependency on other transaction has been found during the read.
-    Dependency(TxnIndex),
-    /// Delta application failed, txn execution should fail.
-    DeltaApplicationFailure,
+///
+/// TODO: separate V into different generic types for data and modules / code (currently
+/// both WriteOp for executor, and use extract_raw_bytes. data for aggregators, and
+/// code for computing the module hash.
+pub struct MVHashMap<K, V: TransactionWrite, X: Executable> {
+    data: VersionedData<K, V>,
+    code: VersionedCode<K, V, X>,
 }
 
-/// Returned as Ok(..) when read successfully from the multi-version data-structure.
-#[derive(Debug, PartialEq, Eq)]
-pub enum MVHashMapOutput<V> {
-    /// Result of resolved delta op, always u128. Unlike with `Version`, we return
-    /// actual data because u128 is cheap to copy amd validation can be done correctly
-    /// on values as well (ABA is not a problem).
-    Resolved(u128),
-    /// Information from the last versioned-write. Note that the version is returned
-    /// and not the data to avoid passing big values around.
-    Version(Version, Arc<V>),
-}
+impl<K: ModulePath + Hash + Clone + Eq, V: TransactionWrite, X: Executable> MVHashMap<K, V, X> {
+    // -----------------------------------
+    // Functions shared for data and code.
 
-impl<K: Hash + Clone + Eq, V: TransactionWrite> MVHashMap<K, V> {
-    pub fn new() -> MVHashMap<K, V> {
+    pub fn new(code_cache: Option<VersionedCode<K, V, X>>) -> MVHashMap<K, V, X> {
         MVHashMap {
-            data: DashMap::new(),
-            delta_keys: Mutex::new(Vec::new()),
+            data: VersionedData::new(),
+            code: code_cache.unwrap_or_default(),
```

Comment:
why do we need the arg code_cache?

Reply:
to re-use between blocks.. I could delete and introduce later, probably will change the API anyway, so I'll get rid of it.

Reply:
added a comment.